### ISSUE TITLE
Incorrect report of self-referencing type for static fields

### ISSUE THREAD

**Issue by boris-kolar** [microsoft/TypeScript#62552]
### 🔎 Search Terms

TS7022
implicitly has type 'any'

### 🕗 Version & Regression Information

- The issue applies to all TypeScript 5 versions and TypeScript 6 nightly

### ⏯ Playground Link

https://www.typescriptlang.org/play/?ts=6.0.0-dev.20251006#code/GYVwdgxgLglg9mABDAJgHgCoD4AUAPALkQwEoiNEBvAKEUQCcBTKEepPagX2uogQGcoiAGJw4iALzIUOCABsAhv35VaiQQtgQGjBSgRyAnomBjJ0nABYATCUQB6e4gCi9enHrEAygHYADNbWXCQ8fGCCiADyANbm8koqNHQaWjp6Bsam4lKoVrYOTgBy4oxuHlzUQA

### 💻 Code

```ts
function id<T>(x: T): T {
  return x
}

const Foo = id(class {
  static readonly foo = id(42) // Error TS7022
})

const Ok = class {
  static readonly foo = id(42) // No error
}
```

### 🙁 Actual behavior

Type error TS7022 reported

### 🙂 Expected behavior

No type error
